WIP: Transfer manager dummy PR

aaronlehmann opened the pull request:
This work adds a transfer manager which deduplicates and schedules transfers, and also an upload manager and download manager that build on top of the transfer manager to provide high-level interfaces for uploads and downloads. The push and pull code is modified to use these building blocks.

Some benefits of the changes:
- Simplification of push/pull code
- Pushes can upload layers concurrently
- Failed downloads and uploads are retried after backoff delays
- Cancellation is supported, but individual transfers will only be cancelled if all pushes or pulls using them are cancelled.
- The distribution code is decoupled from Docker Engine packages and API conventions (i.e. streamformatter), which will make it easier to split out.

This PR is to aggregate the commits for review comments. Do not merge.

Commit:
Add download mananger

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Use RootFS instead of Layer for result

Pass in an initial RootFS to handle base layers.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Replace Layer interface method with DiffID

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Remove Registered from main interface. Detect it by casting.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
godoc

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Concurrency limit

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Use the download manager for v1 pulls

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Remove Pool

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Add upload manager; use it for uploads

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
v1 pull cleanups

- Remove pull -a concurrency
- Remove unnecessary json unmarshal

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Make download manager's Download a blocking function

Change the interface between the puller and the daemon to send progress
over a channel.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Make the upload manager's Upload a blocking function

Change the interface between the pusher and the daemon to send progress
over a channel.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Use ioutils.NewReadCloserWrapper instead of a custom wrapper

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Implement retries

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Implement cancellation

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

stevvooe commented on distribution/pull.go:
```diff
@@ -165,11 +168,11 @@ func Pull(ref reference.Named, imagePullConfig *ImagePullConfig) error {
+		out <- xfer.Progress{Message: "Status: Downloaded newer image for " + requestedTag}
 	} else {
-		out.Write(sf.FormatStatus("", "Status: Image is up to date for %s", requestedTag))
+		out <- xfer.Progress{Message: "Status: Image is up to date for " + requestedTag}
```
Make sure this can never block. I think these are buffered, but you may want to use a select to ensure they don't block.

In distribution/notifications, we have an accept loop ensuring these blocks never happen, with reliable sinks to each listener. Might be good to seek out inspiration there.

aaronlehmann replied:
Yes, these channels are buffered. Currently the buffer size is 100 items.

The problem with doing a non-blocking write is that I don't think we want to lose messages if the client can't keep up. Incremental progress messages are unimportant, but messages like this one should not be lost.

Maybe it would make sense to use non-blocking writes in the progress reader (except when it reaches 100%), but use blocking writes for informational messages like this one.

tonistiigi replied:
I we want we could be smart about the speed. `writeProgress` would never block and store the last written item by ID. Readers could read from that list on their own pace.

aaronlehmann replied:
@tonistiigi I guess in that model, instead of pull/push communicating with the daemon code over a channel, there would be a blocking callback function that daemon could call in a loop to get progress updates? Or are you envisioning it some other way?

tonistiigi replied:
From client side it would be same that I described in https://github.com/aaronlehmann/docker/pull/5#discussion_r45641723 . No preference if it is a callback or a channel, but read and write ends don't have to be same objects. And in callback there could also be a different object that doesn't require ID.

stevvooe replied:
So, would it make sense to have a leaky progress model, where un-sendable updates just get dropped?

aaronlehmann replied:
Check out https://github.com/aaronlehmann/docker/commit/79022b7a6b76d377188c0d86fb0c71ad43d82795. I think this should address the blocking issues nicely. It changes the model so that instead of progress updates synchronously getting sent to all watcher channels, each watcher has a goroutine that waits for updates. Watchers may miss intermediate progress updates if they're too slow to keep up, but it's okay (think of them as "eventually consistent").

stevvooe replied:
@aaronlehmann I'm not going to pretend to understand that code but I think the concept is correct. Basically, you have a broadcast loop, that looks like this:

``` go
for _, dst := range dsts {
    select {
    case dst <- msg:
    default:
        // skip destination
    }
}
```

Listeners must be waiting at the time of send, or the channel must be buffered, to get an update. This can also be accomplished by have a reliable destination (from [notifications](https://github.com/docker/distribution/blob/master/notifications/sinks.go#L48)):

``` go
func (b *Broadcaster) Write(events ...Event) error {
    select {
    case b.events <- events:
    case <-b.closed:
        return ErrSinkClosed
    }
    return nil
}
```

On the other side is a goroutine that ensures the events are pulled from b.events and dispatched to a waiting `Sink`. The target `Sink.Write` function must not block. To provide that guarantee, we back the other side with a queue. We could just as well provide that guarantee by dropping anything that can't be immediately sent.

aaronlehmann replied:
I made some changes to the previous implementation and replaced this commit with https://github.com/aaronlehmann/docker/commit/84a7454ce355620b3759d1a17b1e8bb88bc9683a. The main difference is that the `Release` method was made to block until the progress receiving goroutine for that finishes, to avoid a write-after-close panic. This necessitated being able to terminate an individual goroutine instead of the whole set, so I moved from `sync.Cond` to a signal channel for each goroutine.

stevvooe commented on distribution/xfer/download.go:
```diff
@@ -0,0 +1,299 @@
+}
+
+// A DownloadDescriptor references a layer that may need to be downloaded.
+type DownloadDescriptor interface {
```
Might be better to come up with a different name than "descriptor". This is more like a "job" or "task".

aaronlehmann replied:
Since the download manager decides which of these are already present and which need to be downloaded, I'm not sure "job" or "task" is ideal either. "dependency" comes to mind. What about `DownloadDependency`?

stevvooe replied:
I see. I'm not sure I understand `DownloadDependency`, but aren't we joining a possibly ongoing task or do we have this elsewhere in the model?

aaronlehmann replied:
For each "descriptor" (or whatever we end up calling it), it's possible that we will either:

a) start a download task
b) join an existing download task
c) find that no download is necessary

My issue was "job" or "task" is that it makes it sound like each item will necessarily involve a download, which is not the case in (c). I don't have strong feelings about this though.

stevvooe replied:
It really is like a "task descriptor"...

tonistiigi commented on daemon/daemon.go:
```diff
@@ -1040,23 +1044,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
 
-	return distribution.Pull(ref, imagePullConfig)
+	err := distribution.Pull(ctx, ref, imagePullConfig)
+	close(progressChan)
```
I don't think you should do this, only a sender should close a channel.

aaronlehmann replied:
This lets `writeDistributionProgress` do a for...range loop. Not sure why you think closing the channel here is bad?

tonistiigi replied:
Why can't `Pull` close it instead when it completes?

aaronlehmann replied:
It doesn't seem right to me for `Pull` to close a channel that it didn't create. What if some code wanted to run two run two pulls with the same output channel?

tonistiigi replied:
I think its just one of the best practices as sending on a closed channel panics.

edit: for example here https://tour.golang.org/concurrency/4

stevvooe replied:
The trick here is to make sure the listen side never attempts to close the channel. Patterns like this work fine:

```
select {
case <-closed:
    return ErrClosed
default:
    close(closed)
}
```

The key is ensuring ownership on the channel. Since this code owns the channel and delegates sending and receiving, I don't think we are running afoul of the best practice. We would be in the close was done in `writeDistributionProgress`. We only need to be sure that the channel can only ever be closed once and once it is closed, no sends can occur.

While sending on a closed channel panics, I don't this applies when using a select statement, since the operation cannot proceed. (I may be wrong here).

tonistiigi commented on distribution/pull_v2.go:
```diff
@@ -15,26 +15,23 @@ import (
 )
 
 type v2Puller struct {
+	ctx            context.Context
```
I'm not sure if this is right pattern to use. You should pass a context with the functions you call and make a new ones if you need to split to concurrent goroutines.

aaronlehmann replied:
Tried it both ways, ended up preferring this one for being cleaner and less invasive. There are some other examples of a ctx being part of a struct that carries context in the distribution repo. The pull_v2 code doesn't split into concurrent goroutines so I don't think that's a concern.

stevvooe replied:
> You should pass a context with the functions you call and make a new ones if you need to split to concurrent goroutines.

We definitely break this rule. Typically, we break the rule where all of the function calls will use the same context, such as with a reader or when doing serialization, since we can't change the method signature to include the context.

@aaronlehmann Are we doing this to avoid passing context around with puller calls?

tonistiigi replied:
Yes, for a reader that makes sense.

aaronlehmann replied:
@stevvooe: Yes, it's just cosmetic to avoid passing it to every function (since they all share the puller struct). We can add it to the function signature if that's preferred.

stevvooe replied:
@aaronlehmann I think that's preferred, unless we want to control context crossing, explicitly, which we aren't doing here. Yes, it is ugly, but ultimately, powerful, especially with granular timeout control.

aaronlehmann replied:
Changed the code so contexts are passed as function arguments instead of being stored in the puller structs.

tonistiigi commented on distribution/pull_v2.go:
```diff
@@ -98,91 +93,114 @@ func (p *v2Puller) pullV2Repository(ref reference.Named) (err error) {
-	})
-	io.Copy(di.tmpFile, reader)
+	transferDone := make(chan struct{})
+	go func() {
```
would passing `ctx` to `NewProgeressReader` (or copy) be cleaner?

tonistiigi commented on distribution/pull_v2.go:
```diff
@@ -98,91 +93,114 @@ func (p *v2Puller) pullV2Repository(ref reference.Named) (err error) {
+			return
+		}
+	}()
+	io.Copy(tmpFile, reader)
```
Have to check error here. This error should become true if `ctx` is cancelled.

tonistiigi commented on distribution/xfer/download.go:
```diff
@@ -0,0 +1,299 @@
+
+			d.layer, err = d.layerStore.Register(inflatedLayerData, parentLayer)
+			close(registerDone)
+			if err != nil {
```
No retry here?

aaronlehmann replied:
Do you think it makes sense to retry in this case? There are two common reasons I can think of why `Register` would fail: disk full, or a corrupt tar stream. In either case, retrying the `Register` won't help, and neither will retrying the download, because the checksum was already verified (except in v1, but I want to avoid adding v1-specific behavior).

tonistiigi commented on distribution/pull_v2.go:
```diff
@@ -98,91 +93,114 @@ func (p *v2Puller) pullV2Repository(ref reference.Named) (err error) {
 
-	di.err <- nil
+	tmpFile.Seek(0, 0)
+	return ioutils.NewReadCloserWrapper(tmpFile, tmpFileCloser(tmpFile)), desc.Size, nil
```
What cancels this reader when `ctx` is cancelled?

aaronlehmann replied:
The download manager: https://github.com/aaronlehmann/docker/blob/download-manager-rebased/distribution/xfer/download.go#L227 and https://github.com/aaronlehmann/docker/blob/download-manager-rebased/distribution/xfer/download.go#L259

tonistiigi replied:
But is there anything that would cancel the extraction phase itself. When its blocked on `layerstore.Register`?

aaronlehmann replied:
Doesn't closing `inflatedLayerData` on line 260 accomplish that?

tonistiigi replied:
I guess it does. No idea if its safe though. Normally calling `Read` and `Close` in different goroutines isn't safe.

aaronlehmann replied:
I took a look at these particular types, and it doesn't seem unsafe, but unfortunately `Close` on a gzip reader doesn't seem to actually do anything. I looked into closing `downloadReader` instead, but this is a wrapper around `os.File`, and looking at the standard library code, there's no synchronization to protect `file.fd`, so there are potential weird race conditions where a subsequent read could read from a different file that ends up with the same file descriptor.

I'm going to look into passing a context to `NewProgressReader` and implementing the cancellation there. That would also let other code that uses `ProgressReader` support cancellation in a simpler way. But it's going to require `NewProgressReader` to spawn a goroutine, which means the progress reader will have to have a well-defined lifecycle. I'll give it a try and see what this ends up looking like.

aaronlehmann replied:
I implemented a cancellable progress reader in https://github.com/aaronlehmann/docker/commit/1d9fb7ea14f94877f1328d8133e8923e4bf2110b. Does this look good?

In this same commit, I changed `layer.Layer`'s `TarStream` to return an `io.ReadCloser` so that uploads could be safely interrupted. I'm thinking about taking this part of the commit and making a PR against master, because the current code on master can leak if an upload or export fails partway through. @tonistiigi, what do you think?

aaronlehmann replied:
Implemented pipe-based `Read` interruption for the progress reader in https://github.com/aaronlehmann/docker/commit/360ae6564b3d0633add91a3d755d3e0975a72e30. This commit replaces https://github.com/aaronlehmann/docker/commit/1d9fb7ea14f94877f1328d8133e8923e4bf2110b.

Also cleaned up the `select` statements @tonistiigi didn't like.

Commit:
Change progress broadcasting model

Instead of writing every update to all watcher channels, have a
goroutine for each watcher which is woken up every time there's an
update. Slow watchers may miss intermediate updates, but that's fine -
they will catch up eventually.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Add ioutils.NewCancelReadCloser

This wraps a ReadCloser to make read operations cancellable with a
context.

Use this in combination with ProgressReader to make downloads, uploads,
and layer registration interruptible without extra logic.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Move contexts from structs to function arguments

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
golint fix

Commit:
Avoid redundant downloads of duplicate (empty) layers

When two descriptors have the same key, make the second one wait for the
first one to be registered, and then copy its data out of the layer
store instead of using a separate download.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
@tonistiigi I added https://github.com/aaronlehmann/docker/commit/9fe1187947294a6f10aa5ce9c44f798183cc9e73 which prevents redundant downloads (of empty layers). For each pull, it only starts a download the first time it sees each unique `Key`. If it sees the same key again later in the descriptor list, instead of starting a fresh download, it uses a different transfer function that waits for the first download to complete, and copies its data out of the layer store into the appropriate place in the layer chain.

Commit:
Add unit tests for transfer.go

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>
